File: backend/src/main/java/com/lerneon/backend/services/DepartmentService.java
package com.lerneon.backend.services;

import com.lerneon.backend.models.entity.Department;
import com.lerneon.backend.models.payload.request.DepartmentRequest;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

public interface DepartmentService {
    Page<Department> findAllDepartments(Pageable pageable);

    Department findDepartmentById(Integer id);

    Department createDepartment(DepartmentRequest departmentRequest);

    Department updateDepartment(Integer id, DepartmentRequest departmentRequest);

    Department deleteDepartment(Integer id);
}

File: client/services/department.service.ts
import { PageResponse } from "@/types/payload/response/common/page.response";
import { Department } from "@/types/entity/department";
import { ApiClient } from "@/services/config/api.client";
import { DepartmentRequest } from "@/types/payload/request/department.request";
import { SuccessResponse } from "@/types/payload/response/common/success.response";

export class DepartmentService {
  static async findAllDepartments(): Promise<PageResponse<Department[]>> {
    return await ApiClient.get("department").json<PageResponse<Department[]>>();
  }

  static async findDepartmentById(
    id: number,
  ): Promise<PageResponse<Department>> {
    return await ApiClient.get(`department/${id}`).json<
      PageResponse<Department>
    >();
  }

  static async createDepartment(
    data: DepartmentRequest,
  ): Promise<SuccessResponse<Department>> {
    return ApiClient.post("department", {
      json: data,
    }).json<SuccessResponse<Department>>();
  }

  static async updateDepartment(
    id: number,
    data: DepartmentRequest,
  ): Promise<SuccessResponse<Department>> {
    return ApiClient.put(`department/${id}`, {
      json: data,
    }).json<SuccessResponse<Department>>();
  }

  static async deleteDepartment(
    id: number,
  ): Promise<SuccessResponse<Department>> {
    return ApiClient.delete(`department/${id}`).json<
      SuccessResponse<Department>
    >();
  }
}
